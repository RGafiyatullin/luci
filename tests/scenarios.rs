use insta::{assert_debug_snapshot, assert_yaml_snapshot};
use luci::{
    execution_graph::ExecutionGraph,
    messages::{Messages, Mock},
    scenario::Scenario,
};
use test_case::test_case;

#[test_case("01", true)]
#[test_case("02", true)]
#[test_case("03", true)]
#[test_case("04", false)]
fn run(name: &str, build_graph: bool) {
    let input = format!("tests/scenarios/{name}.yaml");
    let scenario: Scenario =
        serde_yaml::from_str(std::fs::read_to_string(input).expect("read file").as_str())
            .expect("serde::de");
    assert_debug_snapshot!(format!("{name}.debug"), scenario);
    assert_yaml_snapshot!(format!("{name}.yaml"), scenario);

    let messages = Messages::new()
        .with(Mock::msg("one::two::Three"))
        .with(Mock::req("one::two::IsAThree"))
        .with(Mock::msg("protocol_basic::Start"))
        .with(Mock::msg("protocol_basic::Started"))
        .with(Mock::msg("protocol_basic::KeepAlive"));

    if build_graph {
        let _exec_graph = ExecutionGraph::builder(messages)
            .build(&scenario)
            .expect("build exec-graph");
    }
}
