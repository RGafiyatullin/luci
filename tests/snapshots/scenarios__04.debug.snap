---
source: tests/scenarios.rs
expression: scenario
---
Scenario {
    types: [
        TypeAlias {
            type_name: "protocol_basic::KeepAlive",
            type_alias: MessageName(
                "KeepAlive",
            ),
        },
    ],
    sub: [
        SubDef {
            def_path: "04/startup.yaml",
            sub_name: SubName(
                "startup",
            ),
            no_extra: NoExtra,
        },
    ],
    cast: [
        ActorName(
            "a",
        ),
        ActorName(
            "b",
        ),
    ],
    events: [
        EventDef {
            id: EventName(
                "total-test-duration",
            ),
            require: None,
            after: [],
            kind: Delay(
                EventDelay {
                    delay_for: 15s,
                    delay_step: 1s,
                    no_extra: NoExtra,
                },
            ),
            no_extra: NoExtra,
        },
        EventDef {
            id: EventName(
                "standard-startup-routine",
            ),
            require: None,
            after: [],
            kind: Call(
                EventCallSub {
                    sub: SubName(
                        "startup",
                    ),
                    cast: [
                        CallSubBindActor {
                            outer_name: ActorName(
                                "a",
                            ),
                            inner_name: ActorName(
                                "subject",
                            ),
                            no_extra: NoExtra,
                        },
                        CallSubBindActor {
                            outer_name: ActorName(
                                "b",
                            ),
                            inner_name: ActorName(
                                "object",
                            ),
                            no_extra: NoExtra,
                        },
                    ],
                    input: CallSubBindValue {
                        dst: String("$component_id"),
                        src: Literal(
                            String("the-thingy"),
                        ),
                        no_extra: NoExtra,
                    },
                    output: CallSubBindValue {
                        dst: String("$the_thingys_launch_id"),
                        src: Bind(
                            String("$launch_id"),
                        ),
                        no_extra: NoExtra,
                    },
                    no_extra: NoExtra,
                },
            ),
            no_extra: NoExtra,
        },
        EventDef {
            id: EventName(
                "the-thingy-is-alive",
            ),
            require: Some(
                Reached,
            ),
            after: [],
            kind: Recv(
                EventRecv {
                    message_type: MessageName(
                        "KeepAlive",
                    ),
                    message_data: Bind(
                        Object {
                            "launch_id": String("$the_thingys_launch_id"),
                        },
                    ),
                    from: Some(
                        ActorName(
                            "b",
                        ),
                    ),
                    to: Some(
                        ActorName(
                            "a",
                        ),
                    ),
                    no_extra: NoExtra,
                },
            ),
            no_extra: NoExtra,
        },
    ],
    no_extra: NoExtra,
}
